### Importing libraries and reading csv file

In [1]:
import numpy as np 
import pandas as pd 
import sklearn 
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.model_selection import train_test_split # Import train_test_split function 
dataset=pd.read_csv("mnist.csv")

In [ ]:
print(dataset) 

In [13]:
dataset.shape

(42000, 785)

###  features

In [14]:
x=dataset.iloc[:,1:dataset.index[-1]+1].values  #features 
print(x) 

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [15]:
x.shape

(42000, 784)

### Target variable

In [20]:
y=dataset.iloc[:,0].values #target 
print(y)

[1 0 1 ... 7 6 9]


In [21]:
#Spilt dataset into training set and test set 
X_train,X_test, y_train, y_test=train_test_split(x, y, test_size=0.3, random_state=1) # 70% training and 30% test

### KNN method with n=3 with default distance metric

In [22]:
classifier= KNeighborsClassifier(n_neighbors=3) #3
classifier.fit(x,y) #training
classifier.fit(X_train,y_train)
print(X_test)
#print(y_test)
y_predict=classifier.predict(X_test) #testing
print(y_predict)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
[1 5 2 ... 2 3 0]


In [23]:
from sklearn.metrics import confusion_matrix 
confusion_matrix(y_test, y_predict) 

array([[1238,    0,    1,    0,    0,    2,    4,    1,    0,    1],
       [   0, 1391,    1,    0,    0,    0,    1,    1,    0,    1],
       [   3,   13, 1212,    2,    1,    0,    1,   19,    5,    1],
       [   3,    3,   13, 1263,    0,   20,    1,    6,    8,    4],
       [   1,   10,    0,    0, 1187,    0,    0,    1,    0,   33],
       [   6,    2,    1,   15,    1, 1063,   16,    0,    2,    5],
       [   8,    1,    1,    0,    2,    3, 1243,    0,    0,    0],
       [   0,   11,    5,    1,    4,    0,    0, 1246,    0,   18],
       [   8,   22,    9,   17,    7,   17,    8,    4, 1144,   13],
       [   8,    1,    1,    7,   12,    2,    1,   21,    2, 1190]],
      dtype=int64)

In [24]:
#accuracy of predicted labels
from sklearn.metrics import accuracy_score 
accuracy_score(y_test,y_predict) 

0.9664285714285714

### Manhattan distance

In [28]:
classifier= KNeighborsClassifier(n_neighbors=3,metric="manhattan") #3
classifier.fit(x,y) #training
classifier.fit(X_train,y_train)
print(X_test)
#print(y_test)
y_predict=classifier.predict(X_test) #testing
print(y_predict)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
[1 5 2 ... 2 3 0]


In [29]:
confusion_matrix(y_test, y_predict) 

array([[1240,    1,    1,    0,    0,    2,    2,    1,    0,    0],
       [   0, 1391,    1,    0,    0,    0,    1,    1,    0,    1],
       [   7,   20, 1194,    1,    0,    1,    2,   27,    4,    1],
       [   3,    4,   14, 1264,    1,   16,    0,   10,    5,    4],
       [   2,   13,    0,    0, 1174,    0,    2,    5,    0,   36],
       [   4,    4,    2,   23,    3, 1053,   14,    1,    2,    5],
       [   8,    1,    1,    0,    2,    4, 1242,    0,    0,    0],
       [   0,   15,    5,    2,    4,    0,    0, 1244,    0,   15],
       [  10,   28,   11,   23,    9,   28,    8,    5, 1113,   14],
       [   9,    4,    4,   10,   20,    2,    1,   26,    1, 1168]],
      dtype=int64)

In [30]:
accuracy_score(y_test,y_predict) 

0.9589682539682539

In [31]:
classifier= KNeighborsClassifier(n_neighbors=3,metric="chebyshev") #3
classifier.fit(x,y) #training
classifier.fit(X_train,y_train)
print(X_test)
#print(y_test)
y_predict=classifier.predict(X_test) #testing
print(y_predict)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
[1 5 2 ... 2 3 0]


In [32]:
accuracy_score(y_test,y_predict) 

0.7673015873015873

In [36]:
classifier= KNeighborsClassifier(n_neighbors=5) #3
classifier.fit(x,y) #training
classifier.fit(X_train,y_train)
print(X_test)
#print(y_test)
y_predict=classifier.predict(X_test) #testing
print(y_predict)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
[1 5 2 ... 2 3 0]


In [37]:
accuracy_score(y_test,y_predict) 

0.966031746031746

### Default distance metric gives higher accuracy than others.Though this accuracy can be improved by various methods like using CNN.KNN gives a accuracy of about 96%.